platform/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/camera/GestureDetectorActivity.java

```diff
@@ -0,0 +1,422 @@
+      case R.id.menu_gesture_none:
+        return true;
+      case R.id.menu_gesture_focus_point:
+        focalPointLatLng = new LatLng(51.50325, -0.12968);
+        marker = mapboxMap.addMarker(new MarkerOptions().position(focalPointLatLng));
+        mapboxMap.easeCamera(CameraUpdateFactory.newLatLngZoom(focalPointLatLng, 16));
+        mapboxMap.getUiSettings().setFocalPoint(mapboxMap.getProjection().toScreenLocation(focalPointLatLng));
+        return true;
+      case R.id.menu_gesture_animation:
+        animationsEnabled = !animationsEnabled;
+        mapboxMap.getUiSettings().setAllVelocityAnimationsEnabled(animationsEnabled);
+    }
+    return super.onOptionsItemSelected(item);
+  }
+
+  private void resetModes() {
+    focalPointLatLng = null;
+    mapboxMap.getUiSettings().setFocalPoint(null);
+
+    gesturesManager.getMoveGestureDetector().setMoveThreshold(
+      gesturesManager.getMoveGestureDetector().getDefaultMoveThreshold()
+    );
+
+    if (marker != null) {
+      mapboxMap.removeMarker(marker);
+      marker = null;
+    }
+  }
+
+  private void recalculateFocalPoint() {
+    if (focalPointLatLng != null) {
+      mapboxMap.getUiSettings().setFocalPoint(
+        mapboxMap.getProjection().toScreenLocation(focalPointLatLng)
+      );
+    }
+  }
+
+  private static class GestureAlertsAdapter extends RecyclerView.Adapter<GestureAlertsAdapter.ViewHolder> {
+
+    private boolean isUpdating;
+    private final Handler updateHandler = new Handler();
+    private final List<GestureAlert> alerts = new ArrayList<>();
+
+    public static class ViewHolder extends RecyclerView.ViewHolder {
+
+      TextView alertMessageTv;
+
+      @ColorInt
+      public int textColor;
+
+      ViewHolder(View view) {
+        super(view);
+        Typeface typeface = FontCache.get("Roboto-Regular.ttf", view.getContext());
+        alertMessageTv = (TextView) view.findViewById(R.id.alert_message);
+        alertMessageTv.setTypeface(typeface);
+      }
+    }
+
+
+    @Override
+    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
+      View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_gesture_alert, parent, false);
+      return new ViewHolder(view);
+    }
+
+    @Override
+    public void onBindViewHolder(ViewHolder holder, int position) {
+      GestureAlert alert = alerts.get(position);
+      holder.alertMessageTv.setText(alert.getMessage());
+      holder.alertMessageTv.setTextColor(
+        ContextCompat.getColor(holder.alertMessageTv.getContext(), alert.getColor()));
+    }
+
+    @Override
+    public int getItemCount() {
+      return alerts.size();
+    }
+
+    void addAlert(GestureAlert alert) {
+      for (GestureAlert gestureAlert : alerts) {
+        if (gestureAlert.getAlertType() != GestureAlert.TYPE_PROGRESS) {
+          break;
+        }
+
+        if (alert.getAlertType() == GestureAlert.TYPE_PROGRESS && gestureAlert.equals(alert)) {
+          return;
+        }
+      }
```

Comment:
not actionable but flagging that code for loops as above with inclusion of break/return increases [cyclomatic complexity](https://en.wikipedia.org/wiki/Cyclomatic_complexity). This makes reading the code more difficult as it makes predicting the outcome harder.

Reply:
Interesting, thanks for pointing this out.